src/interpolate.jl

```diff
@@ -40,6 +40,113 @@ function cell2node(cell_centered_data::AbstractArray{Float64})
+
+  # Get node coordinates for DG locations on reference element
+  nodes_in, _ = gauss_lobatto_nodes_weights(n_nodes_in)
+
+  # New unstructured data has one dimension less.
+  # The redundant element ids are removed later.
+  new_unstructured_data = similar(unstructured_data[1, ..])
+
+  # Declare new empty arrays to fill in new coordinates and levels
+  new_coordinates = Array{Float64}(undef, 2, 0)
+  new_levels = Array{eltype(levels)}(undef, 0)
+
+  # Counter for new element ids
+  new_id = 0
+
+  # Save vandermonde matrices in a Dict to prevent redundant generation
+  vandermonde_to_2d = Dict()
+
+  # Limits of domain in slice_axis dimension
+  lower_limit = center_level_0[slice_axis_dimension] - length_level_0 / 2
+  upper_limit = center_level_0[slice_axis_dimension] + length_level_0 / 2
+
+  if slice_axis_intersect < lower_limit || slice_axis_intersect > upper_limit
+    error("slice_axis_intersect $slice_axis_intersect outside of domain")
+  end
+
+  for element_id in 1:n_elements
+    # Distance from center to border of this element (half the length)
+    element_length = length_level_0 / 2^levels[element_id]
+    first_coordinate = coordinates[:, element_id] .- element_length / 2
+    last_coordinate = coordinates[:, element_id] .+ element_length / 2
+
+    # Check if slice plane and current element intersect
+    # The upper limit check is needed because of the > in the first check
+    if (first_coordinate[slice_axis_dimension] <= slice_axis_intersect &&
+          last_coordinate[slice_axis_dimension] > slice_axis_intersect) ||
+        (slice_axis_intersect == upper_limit &&
+          last_coordinate[slice_axis_dimension] == upper_limit)
+      # This element is of interest
+      new_id += 1
+
+      # Add element to new coordinates and levels
+      new_coordinates = hcat(new_coordinates, coordinates[other_dimensions, element_id])
+      push!(new_levels, levels[element_id])
+
+      # Construct vandermonde matrix (or load from Dict if possible)
+      normalized_intersect =
+          (slice_axis_intersect - first_coordinate[slice_axis_dimension]) /
+          element_length * 2 - 1
+
+      if haskey(vandermonde_to_2d, normalized_intersect)
+        vandermonde = vandermonde_to_2d[normalized_intersect]
+      else
+        # Generate vandermonde matrix to interpolate values at nodes_in to one value
+        vandermonde = polynomial_interpolation_matrix(nodes_in, [normalized_intersect])
+        vandermonde_to_2d[normalized_intersect] = vandermonde
+      end
+
+      # 1D interpolation to specified slice plane
+      for i in 1:n_nodes_in
+        for ii in 1:n_nodes_in
+          if slice_axis == :x
+            data = unstructured_data[:, i, ii, element_id, :]
+          elseif slice_axis == :y
+            data = unstructured_data[i, :, ii, element_id, :]
+          elseif slice_axis == :z
+            data = unstructured_data[i, ii, :, element_id, :]
+          end
```

Comment:
Use `@views` to avoid copies:
```suggestion
          if slice_axis == :x
            @views data = unstructured_data[:, i, ii, element_id, :]
          elseif slice_axis == :y
            @views data = unstructured_data[i, :, ii, element_id, :]
          elseif slice_axis == :z
            @views data = unstructured_data[i, ii, :, element_id, :]
          end
```

Reply:
OTOH, the interpolation might be faster if working on contiguous data... maybe time it and then decide which version to use...

Reply:
I quickly tested it and it seems like the version without `@views` is about twice as fast. Why is that the case? How can access to `@views` be slower?

Reply:
You could have a look at `@code_warntype`:
```julia
julia> A = rand(4,4,4,4);

julia> function foo(unstructured_data, slice_axis)
           i = 1
           ii = 1
           element_id = 1
           if slice_axis == :x
               @views data = unstructured_data[:, i, ii, element_id, :]
           elseif slice_axis == :y
               @views data = unstructured_data[i, :, ii, element_id, :]
           elseif slice_axis == :z
               @views data = unstructured_data[i, ii, :, element_id, :]
           end
           sum(data)
       end
foo (generic function with 1 method)

julia> @code_warntype foo(A, :x)
Variables
  #self#::Core.Compiler.Const(foo, false)
  unstructured_data::Array{Float64,4}
  slice_axis::Symbol
  i::Int64
  ii::Int64
  element_id::Int64
  data::SubArray
```
That's type instability.

Reply:
Can you explain why taking a sub-array leads to type instability here?

Reply:
Because the indices are type parameters of the created `SubArray`:
```julia
julia> A = rand(4, 4, 4);

julia> view(A, :, 1, :) |> typeof
SubArray{Float64,2,Array{Float64,3},Tuple{Base.Slice{Base.OneTo{Int64}},Int64,Base.Slice{Base.OneTo{Int64}}},false}

julia> view(A, :, :, 1) |> typeof
SubArray{Float64,2,Array{Float64,3},Tuple{Base.Slice{Base.OneTo{Int64}},Base.Slice{Base.OneTo{Int64}},Int64},true}
```

Reply:
But why does Julia not automatically optimize `interpolate_nodes` for different types of `data_in`?

Reply:
I'm not sure what you mean exactly. Julia will specialize on the type of `unstructured_data` and also on `data` *after it has been created*. However, the code is not type-stable, so the concrete type of `data` is only available at runtime. Hence, dynamic dispatch at runtime is necessary here.

You could avoid that inside this function by making the choice depending on `slice_axis` type-stable, e.g. by using `slice_axis == Val{:x}()` instead of `slice_axis == :x` so that all type information can be determined at compile time (you should check whether that really happens if you use this approach). Or you just keep the current approach...

Reply:
Okay, I still have a lot to learn about type stability in Julia. I will just keep the current approach here.

Reply:
Another alternative: You could move everything accessing `data` into the `if` blocks to make it (locally) type-stable when using `view`s.